src/traits.rs

```diff
@@ -0,0 +1,151 @@
+use core::fmt::Debug;
+
+/// The argument type. This trait is implemented for types like [`&str`]
+/// and [`&[u8]`], and allows them to be understood by `getargs` enough
+/// to parse them - `getargs` is entirely generic over the type of its
+/// arguments.
+pub trait Argument: Copy + Eq + Debug {
+    /// The short-flag type. For [`&str`], this is [`char`]. For
+    /// [`&[u8]`], this is `u8`.
+    type ShortOpt: Copy + Eq + Debug;
+
+    /// Returns true if this argument signals that no additional options
+    /// should be parsed. This argument will not be included in the
+    /// options or positional arguments. The literal string `--` should
+    /// signal the end of the options, but not if it is followed by any
+    /// more content (which would make it a long option).
+    fn ends_opts(self) -> bool;
+
+    /// Returns a long option if one is present, along with its
+    /// optional value. A long option and its value are the `option` and
+    /// `value` parts of the following string: `--option=value`. The
+    /// value may be empty if `=` is present. Do not include the `--`.
+    fn long_opt(self) -> Option<(Self, Option<Self>)>;
+
+    /// Returns a "short option cluster" if one is present. A "short
+    /// option cluster" is defined as at least one
+    /// [`ShortFlag`][Self::ShortFlag] following a `-` (but not `--`).
+    /// Do not include the `-`.
+    ///
+    /// The given "short option cluster" must be valid for at least one
+    /// [`consume_short_opt`].
+    fn short_cluster(self) -> Option<Self>;
+
+    /// Attempts to consume one short option from a "short option
+    /// cluster". Returns the short option that was consumed (if any)
+    /// and the rest of the cluster (if non-empty).
+    fn consume_short_opt(self) -> (Self::ShortOpt, Option<Self>);
```

Comment:
The idea is that, if `Some`, the return value of this method must always be valid for `consume_short_opt`, but not necessarily `consume_short_val`. That's why `consume_short_val` returns an `Option`. However, it can be `None` to signal that this is the last short option of the cluster, and therefore not valid for `consume_short_opt` *or* `consume_short_val`.

Reply:
> The idea is that, if `Some`, the return value of this method must always be valid for `consume_short_opt`, but not necessarily `consume_short_val`.

When will it be valid for `consume_short_opt` but not `consume_short_val`? If there are more characters left then it should be valid for both, right?

Reply:
> > The idea is that, if `Some`, the return value of this method must always be valid for `consume_short_opt`, but not necessarily `consume_short_val`.
> 
> When will it be valid for `consume_short_opt` but not `consume_short_val`? If there are more characters left then it should be valid for both, right?

I've resolved this concern in the latest commit along with the new docs, the method contracts should be clearer now. :)